### PR TITLE
CMP #10: Add ClassicalShadow MP class

### PR DESCRIPTION
**Pull request by AlbertMitjans**
- Add `CustomMeasurement` class, which takes a tape and a device and returns the measurement value.
- Change name from `ShadowMeasurementProcess` to `ClassicalShadow`, to be more consistent with the function `qml.classical_shadow`
- Add class `_ShadowExpval`, which contains the measurement process related to `qml.shadow_expval`
- Fix `qml.equal` to work with this new class.
- Deprecate `seed_recipes` in `qml.classical_shadow` and `qml.shadow_expval`

### CODE REVIEW

File: pennylane/measurements/classical_shadow.py [hunk 1]
```diff
@@ -274,15 +318,69 @@ def shape(self, device=None):
...
+        Please refer to :func:`~.pennylane.shadow_expval` for detailed documentation.
+
+        Args:
+            tape (QuantumScript): the tape to be processed
+            device (Device): the device used to process the tape
+
+        Returns:
+            float: expectation value estimate.
+        """
+        bits, recipes = qml.classical_shadow(wires=self.wires, seed_recipes=self.seed).process(
```
Comment (eddddddy) [inline pennylane/measurements/classical_shadow.py:365]: The `seed_recipes` argument is just a boolean that tells `qml.classical_shadow` whether to generate a random seed to use for random recipe generation (this makes sure that the same recipes are used when the tape is copied, but the samples can still vary). At the moment you can't actually pass in a manual seed yourself. I think there are two options here:

1. Pass `seed_recipes=self.seed is not None` to `qml.classical_shadow`.
2. Allow `seed_recipes` to take integer values in addition to booleans to let the user manually specify a seed.

File: pennylane/measurements/classical_shadow.py [hunk 2]
```diff
@@ -274,15 +318,69 @@ def shape(self, device=None):
...
+            tape, device
+        )
+        shadow = qml.shadows.ClassicalShadow(bits, recipes, wire_map=self.wires.tolist())
+        return shadow.expval(self.H, self.k)
+
+    @property
+    def numeric_type(self):
+        return float
+
+    def shape(self, device=None):
```
Comment (eddddddy) [inline pennylane/measurements/classical_shadow.py:375]: Docstrings

File: pennylane/measurements/classical_shadow.py [hunk 3]
```diff
@@ -291,19 +389,20 @@ def wires(self):
...
 
     def __copy__(self):
-        obj = super().__copy__()
-        obj.seed = self.seed
-        obj.H = self.H
-        obj.k = self.k
-        return obj
+        return self.__class__(
+            self.return_type,
+            H=self.H,
```
Comment (eddddddy) [inline pennylane/measurements/classical_shadow.py:402]: I think we should copy all the observables in `H` too. At the moment there are no issues as it is, but it would be good to copy just in case.

File: pennylane/ops/functions/equal.py [hunk 4]
```diff
@@ -164,23 +167,32 @@ def _equal_measurements(op1: MeasurementProcess, op2: MeasurementProcess, **kwar
...
+    """Determine whether two MeasurementProcess objects are equal"""
+    eq_m = _equal_measurements(op1, op2)
+    log_base_match = op1.log_base == op2.log_base
+    return eq_m and log_base_match
+
+
+@_equal.register
+# pylint: disable=unused-argument
+def _equal_shadow_measurements(op1: _ShadowExpval, op2: _ShadowExpval, **kwargs):
+    """Determine whether two ClassicalShadow objects are equal"""
```
Comment (eddddddy) [inline pennylane/ops/functions/equal.py:195]: Do we need one for `ClassicalShadow` as well?

File: tests/shadow/test_shadow_measurement.py [hunk 5]
```diff
@@ -25,13 +26,7 @@ def get_circuit(wires, shots, seed_recipes, interface="autograd", device="defaul
...
         and performs the classical shadow measurement
     """
-    if device is not None:
-        dev = qml.device(device, wires=wires, shots=shots)
-    else:
-        dev = qml.device("default.qubit", wires=wires, shots=shots)
-
-        # make the device call the superclass method to switch between the general qubit device and device specific implementations (i.e. for default qubit)
-        dev.classical_shadow = super(type(dev), dev).classical_shadow
+    dev = qml.device(device, wires=wires, shots=shots)
```
Comment (eddddddy) [inline tests/shadow/test_shadow_measurement.py:29]: I think we still have to create a mock device without a classical shadow implementation, instead of relying on the fact that `default.mixed` doesn't have one.

Eventually we'll want to support a fast implementation of classical shadows for `default.mixed`, just like we have for `default.qubit` currently. In this case, the default implementation of classical shadows in `ClassicalShadow.process()` won't be tested and we'll lose coverage.

File: tests/shadow/test_shadow_measurement.py [hunk 6]
```diff
@@ -49,19 +44,13 @@ def get_x_basis_circuit(wires, shots, interface="autograd", device="default.qubi
...
 
     @qml.qnode(dev, interface=interface)
     def circuit():
         for wire in range(wires):
             qml.Hadamard(wire)
-        return qml.classical_shadow(wires=range(wires))
+        return qml.classical_shadow(wires=range(wires), seed_recipes=False)
```
Comment (eddddddy) [inline tests/shadow/test_shadow_measurement.py:53]: Why was this change necessary?

File: tests/shadow/test_shadow_measurement.py [hunk 7]
```diff
@@ -156,7 +133,7 @@ def test_measurement_process_copy(self, wires, seed):
         res = qml.classical_shadow(wires=range(wires), seed_recipes=seed)
 
         copied_res = copy.copy(res)
-        assert type(copied_res) == type(res)
+        assert isinstance(copied_res, type(res))
```
Comment (eddddddy) [inline tests/shadow/test_shadow_measurement.py:136]: Should we also check vice-versa?

File: tests/shadow/test_shadow_measurement.py [hunk 8]
```diff
@@ -165,7 +142,7 @@ def test_measurement_process_copy(self, wires, seed):
     @pytest.mark.parametrize("shots", shots_list)
     @pytest.mark.parametrize("seed", seed_recipes_list)
     @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
-    @pytest.mark.parametrize("device", ["default.qubit", "default.mixed", None])
+    @pytest.mark.parametrize("device", ["default.qubit", "default.mixed"])
```
Comment (eddddddy) [inline tests/shadow/test_shadow_measurement.py:145]: See my previous comment about testing the default implementation.

File: tests/shadow/test_shadow_measurement.py (hunk 5 above)
Comment (AlbertMitjans) [reply tests/shadow/test_shadow_measurement.py]: Makes sense. Changed back to original.

File: tests/shadow/test_shadow_measurement.py (hunk 8 above)
Comment (AlbertMitjans) [reply tests/shadow/test_shadow_measurement.py]: Changed back to original

File: tests/shadow/test_shadow_measurement.py (hunk 6 above)
Comment (AlbertMitjans) [reply tests/shadow/test_shadow_measurement.py]: I don't think it was necessary. I was testing some stuff and left it by mistake. Removed 👍

File: pennylane/ops/functions/equal.py (hunk 4 above)
Comment (AlbertMitjans) [reply pennylane/ops/functions/equal.py:264]: The `ClassicalShadow` class only has one extra argument that the `MeasurementProcess` class, which is `self.seed`. However the seed wasn't used in `qml.equal`, thus we can use the `MeasurementProcess` branch to test it.

File: pennylane/measurements/classical_shadow.py [hunk 9]
```diff
@@ -219,50 +220,93 @@ class ShadowMeasurementProcess(MeasurementProcess):
...
+
+                outcomes[t] = device.generate_samples()[0][mapped_wires]
+
+        return qml.math.cast(qml.math.stack([outcomes, recipes]), dtype=np.int8)
 
+    @property
+    def numeric_type(self):
+        return int
+
+    def shape(self, device=None):
```
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/measurements/classical_shadow.py:309]: The `MeasurementProcess` parent class already has a docstring defined for these methods.

File: pennylane/measurements/classical_shadow.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/measurements/classical_shadow.py]: Good catch!

File: pennylane/measurements/classical_shadow.py (hunk 1 above)
Comment (AlbertMitjans) [reply pennylane/measurements/classical_shadow.py]: Changed `seed_recipes` to `seed`. Now it is an integer instead of a boolean.

File: tests/shadow/test_shadow_measurement.py (hunk 7 above)
Comment (AlbertMitjans) [reply tests/shadow/test_shadow_measurement.py:160]: Changed this to:
```python
assert isinstance(copied_res, ClassicalShadow)
```

File: pennylane/measurements/classical_shadow.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/measurements/classical_shadow.py:304]: Fair.

File: pennylane/_qubit_device.py [hunk 10]
```diff
@@ -1269,13 +1269,10 @@ def classical_shadow(self, obs, circuit):
...
-                # seed the random measurement generation so that recipes
-                # are the same for different executions with the same seed
-                rng = np.random.RandomState(seed)
-                recipes = rng.randint(0, 3, size=(n_snapshots, n_qubits))
-            else:
-                recipes = np.random.randint(0, 3, size=(n_snapshots, n_qubits))
+            # seed the random measurement generation so that recipes
+            # are the same for different executions with the same seed
+            rng = np.random.RandomState(seed)
+            recipes = rng.randint(0, 3, size=(n_snapshots, n_qubits))
```
Comment (eddddddy) [inline pennylane/_qubit_device.py:1275]: 💯

File: pennylane/measurements/classical_shadow.py [hunk 11]
```diff
@@ -219,50 +218,90 @@ class ShadowMeasurementProcess(MeasurementProcess):
...
+        Pauli measurements have shape ``(T, n)``.
+
+        This implementation leverages vectorization and offers a significant speed-up over
+        the generic implementation.
+
+        .. Note::
+
+            This method internally calls ``np.einsum`` which supports at most 52 indices,
+            thus the classical shadow measurement for this device supports at most 52
+            qubits.
```
Comment (eddddddy) [inline pennylane/measurements/classical_shadow.py:254]: This looks like it's copied from the `default.qubit` docstring instead of the generic `_qubit_device` docstring.

File: pennylane/measurements/classical_shadow.py [hunk 12]
```diff
@@ -15,17 +15,18 @@
...
 
 import pennylane as qml
 from pennylane.wires import Wires
 
-from .measurements import MeasurementProcess, MeasurementShapeError, Shadow, ShadowExpval
+from .measurements import CustomMeasurement, MeasurementShapeError, Shadow, ShadowExpval
 
 
-def shadow_expval(H, k=1, seed_recipes=True):
+def shadow_expval(H, k=1, seed=None):
```
Comment (eddddddy) [inline pennylane/measurements/classical_shadow.py:29]: Let's add this to the changelog since I would guess that it counts as a breaking change.

**Review by eddddddy — APPROVED** [review]
Thanks @AlbertMitjans! A couple of last suggestions but otherwise looks good 🔥

File: pennylane/measurements/classical_shadow.py (hunk 11 above)
Comment (AlbertMitjans) [reply pennylane/measurements/classical_shadow.py]: Fixed!

File: pennylane/measurements/classical_shadow.py (hunk 12 above)
Comment (AlbertMitjans) [reply pennylane/measurements/classical_shadow.py]: Done!

**Comment by AlbertMitjans** [comment]
Hi @eddddddy, I added a deprecation cycle for the breaking change in `qml.classical_shadow` and `qml.shadow_expval`. Can you quickly review it?

File: doc/development/deprecations.rst [hunk 13]
```diff
@@ -6,6 +6,13 @@ Deprecations
 Pending deprecations
 --------------------
 
+* The ``seed_recipes`` argument in ``qml.classical_shadow`` and ``qml.shadow_expval`` is deprecated.
+  A new argument ``seed`` has been added, which defaults to None and can contain an integer with the
```
Comment (eddddddy) [inline doc/development/deprecations.rst:10]:
```suggestion
  A new argument ``seed`` has been added, which defaults to ``None`` and can contain an integer with the 
```

**Review by eddddddy — APPROVED** [review]
Looks good @AlbertMitjans 💯

File: doc/releases/changelog-dev.md [hunk 14]
```diff
@@ -331,6 +348,11 @@ Deprecations cycles are tracked at [doc/developement/deprecations.rst](https://d
 * `qml.transforms.measurement_grouping` has been deprecated. Use `qml.transforms.hamiltonian_expand` instead.
   [(#3417)](https://github.com/PennyLaneAI/pennylane/pull/3417)
 
+* The ``seed_recipes`` argument in ``qml.classical_shadow`` and ``qml.shadow_expval`` is deprecated.
```
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md:351]: Done!
